### PR TITLE
ARROW-15672: [C++] Enable CSV writer to control the field delimiter

### PR DESCRIPTION
**Pull request by sanjibansg**
This PR modifies the WriteOptions for CSV to allow changing the default delimiter.

### CODE REVIEW

File: python/pyarrow/_csv.pyx [hunk 1]
```diff
@@ -1025,17 +1025,21 @@ cdef class WriteOptions(_Weakrefable):
     batch_size : int, optional (default 1024)
         How many rows to process together when converting and writing
         CSV data
+    delimiter : char, optional (default `'`)
```
Comment (lidavidm) [inline python/pyarrow/_csv.pyx:1028]: nit, but I think Python would use `"'"`, and char is not a type in Python (it should be str). And in any case, is the delimiter not `,`?

Comment (lidavidm) [reply python/pyarrow/_csv.pyx:1028]: See https://github.com/apache/arrow/blob/36a76914cd42606be7ef847d500af67c64ccef26/python/pyarrow/_csv.pyx#L278 for instance

File: cpp/src/arrow/csv/writer.cc [hunk 2]
```diff
@@ -406,7 +405,7 @@ class CSVWriterImpl : public ipc::RecordBatchWriter {
     std::vector<std::unique_ptr<ColumnPopulator>> populators(schema->num_fields());
     for (int col = 0; col < schema->num_fields(); col++) {
       const std::string& end_chars =
-          col < schema->num_fields() - 1 ? kStrComma : options.eol;
+          col < schema->num_fields() - 1 ? std::string(1,options.delimiter) : options.eol;
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer.cc:408]: While the compiler will likely do this for us, we should probably hoist the string construction outside of the loop.

**Review by lidavidm — COMMENTED** [review]
Thanks for doing this. We should also add unit tests (see https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/writer_test.cc and https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_csv.py) and you'll want to reformat the code here (see https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci)

File: python/pyarrow/_csv.pyx (hunk 1 above)
Comment (sanjibansg) [reply python/pyarrow/_csv.pyx]: Yes, sorry I missed it. Corrected that now.

File: cpp/src/arrow/csv/writer.cc (hunk 2 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer.cc]: Yes, right, that's better.

**Comment by sanjibansg** [comment]
> Thanks for doing this. We should also add unit tests (see https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/writer_test.cc and https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_csv.py) and you'll want to reformat the code here (see https://arrow.apache.org/docs/developers/cpp/development.html#code-style-linting-and-ci)

Adding the tests now,

File: cpp/src/arrow/csv/writer_test.cc [hunk 3]
```diff
@@ -233,7 +241,15 @@ std::vector<WriterTestParams> GenerateTestCases() {
       reject_structural_params({"0123456789", nullptr, "abcde,", nullptr}, "abcde,"),
       reject_structural_params({"0123456789", nullptr, "abcdef,", nullptr}, "abcdef,"),
       reject_structural_params({nullptr, nullptr, ",0123456789", "abcde"}, ",0123456789"),
-      reject_structural_params({"0123456", nullptr, "7\\\"89", ",abcdef"}, "7\"89")};
+      reject_structural_params({"0123456", nullptr, "7\\\"89", ",abcdef"}, "7\"89"),
+      {schema_custom_delimiter, batch_custom_delimiter,
```
Comment (cyb70289) [inline cpp/src/arrow/csv/writer_test.cc:245]: Please insert a comment, e.g., `// exercise custom delimiter`, to separate from simd tests above.

File: python/pyarrow/includes/libarrow.pxd [hunk 4]
```diff
@@ -1713,6 +1713,7 @@ cdef extern from "arrow/csv/api.h" namespace "arrow::csv" nogil:
     cdef cppclass CCSVWriteOptions" arrow::csv::WriteOptions":
         c_bool include_header
         int32_t batch_size
+        unsigned char delimiter
```
Comment (cyb70289) [inline python/pyarrow/includes/libarrow.pxd:1716]: `unsigned char` or `char`?

File: cpp/src/arrow/csv/options.h [hunk 5]
```diff
@@ -194,6 +194,9 @@ struct ARROW_EXPORT WriteOptions {
   /// This number can impact performance.
   int32_t batch_size = 1024;
 
+  /// Field delimiter
+  char delimiter = ',';
```
Comment (cyb70289) [inline cpp/src/arrow/csv/options.h:198]: Shall we treat the customized delimiter as non structural chars in non-quoted string checking [1]? Currently, it's hard coded to check `,` and `\r`, `\n`, `"`.

https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/writer.cc#L213

File: cpp/src/arrow/csv/options.cc [hunk 6]
```diff
@@ -73,6 +73,9 @@ Status ReadOptions::Validate() const {
 WriteOptions WriteOptions::Defaults() { return WriteOptions(); }
 
 Status WriteOptions::Validate() const {
+  if (ARROW_PREDICT_FALSE(delimiter == '\n' || delimiter == '\r')) {
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply cpp/src/arrow/csv/options.cc]: Made the change.

File: cpp/src/arrow/csv/writer_test.cc (hunk 3 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer_test.cc:246]: Made the change.

File: cpp/src/arrow/csv/options.h (hunk 5 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/options.h:198]: I am not sure about its implementation, could you please help me with it.

cc: @westonpace @lidavidm

File: python/pyarrow/includes/libarrow.pxd (hunk 4 above)
Comment (sanjibansg) [reply python/pyarrow/includes/libarrow.pxd]: Made the change.

File: cpp/src/arrow/csv/options.h (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/options.h:198]: No worry, I will upload the change above your PR.

Comment (sanjibansg) [reply cpp/src/arrow/csv/options.h:198]: Thanks!

File: python/pyarrow/tests/test_csv.py [hunk 7]
```diff
@@ -326,7 +326,7 @@ def test_write_options():
     opts = cls()
 
     check_options_class(
-        cls, include_header=[True, False])
+        cls, include_header=[True, False], delimiter=[",", "\t", "|"])
```
Comment (lidavidm) [inline python/pyarrow/tests/test_csv.py:329]: Can we add a case to https://github.com/apache/arrow/blob/db2401c53f63643b06cab4c683cc086b915ec5c4/python/pyarrow/tests/test_csv.py#L1875-L1880 ?

File: cpp/src/arrow/csv/writer_test.cc [hunk 8]
```diff
@@ -233,7 +241,16 @@ std::vector<WriterTestParams> GenerateTestCases() {
...
+      reject_structural_params({"0123456", nullptr, "7\\\"89", ",abcdef"}, "7\"89"),
+      // exercise custom delimiter
+      {schema_custom_delimiter, batch_custom_delimiter,
+       DefaultTestOptions(/*include_header=*/false, /*null_string=*/"",
+                          QuotingStyle::Needed, "\n", /*delimiter=*/'\t'),
+       expected_output_delimiter_tabs},
+      {schema_custom_delimiter, batch_custom_delimiter,
+       DefaultTestOptions(/*include_header=*/false, /*null_string=*/"",
+                          QuotingStyle::Needed, "\n", /*delimiter=*/'|'),
+       expected_output_delimiter_pipe}};
```
Comment (lidavidm) [inline cpp/src/arrow/csv/writer_test.cc:253]: Can we add a couple cases with invalid parameters to make sure they're properly rejected? See how reject_structural_params sets things up.

File: cpp/src/arrow/csv/options.cc [hunk 9]
```diff
@@ -73,6 +73,9 @@ Status ReadOptions::Validate() const {
 WriteOptions WriteOptions::Defaults() { return WriteOptions(); }
 
 Status WriteOptions::Validate() const {
+  if (ARROW_PREDICT_FALSE(delimiter == '\n' || delimiter == '\r' || delimiter == '"')) {
+    return Status::Invalid("WriteOptions: delimiter cannot be \\r or \\n");
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/csv/options.cc:77]: We support customized eol strings excepts `\n, \r, \r\n`, such as `_EOL_` as line separator in csv writer [1].
Is it useful in practice? Our csv parser doesn't suppot it, AFAIK, @pitrou 

Pandas.read_csv can specify a **single char** eol breaker by `lineterminator` parameter [2], works as expected per my test. But looks our eol string is too much?

[1] https://github.com/apache/arrow/blob/master/cpp/src/arrow/csv/writer_test.cc#L116
[2] https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html

File: cpp/src/arrow/csv/options.h (hunk 5 above)
Comment (cyb70289) [reply cpp/src/arrow/csv/options.h:198]: [to-sanjiban.diff.gz](https://github.com/apache/arrow/files/8094142/to-sanjiban.diff.gz)
@sanjibansg , you can apply above diff to your code. Please note this is a quick fix, needs adding test cases, or refinement maybe, and may have format issue.

File: cpp/src/arrow/csv/options.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (sanjibansg) [reply cpp/src/arrow/csv/options.cc]: Made the change for checking with eol as well.

File: cpp/src/arrow/csv/options.h (hunk 5 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/options.h]: okay thanks applying this diff. Thanks a lot @cyb70289!

File: python/pyarrow/tests/test_csv.py (hunk 7 above)
Comment (sanjibansg) [reply python/pyarrow/tests/test_csv.py]: made the change.

File: cpp/src/arrow/csv/writer_test.cc (hunk 8 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer_test.cc]: Made the changes.

File: cpp/src/arrow/csv/writer_test.cc [hunk 10]
```diff
@@ -189,6 +198,12 @@ std::vector<WriterTestParams> GenerateTestCases() {
             /*expected_output*/ "", expected_status_no_quotes_with_structural(error_val)};
   };
 
+  // Schema/expected message for delimiter test
+  auto schema_custom_delimiter = schema({field("a", int64()), field("b", int64())});
+  auto batch_custom_delimiter = R"([{"a": 42, "b": -12}])";
+  auto expected_output_delimiter_tabs = "42\t-12\n";
+  auto expected_output_delimiter_pipe = "42|-12\n";
```
Comment (cyb70289) [inline cpp/src/arrow/csv/writer_test.cc:205]: Looks duplicated, one test should be enough.

File: cpp/src/arrow/csv/writer_test.cc [hunk 11]
```diff
@@ -171,6 +173,13 @@ std::vector<WriterTestParams> GenerateTestCases() {
         "style is \"None\". See RFC4180. Invalid value: ",
         value);
   };
+
+  auto expected_status_illegal_delimiter = [](const char value) {
+    return Status::Invalid(
+        "WriteOptions: delimiter cannot be \\r or \\n or \" or EOL. Invalid value: ",
+        value);
+  };
```
Comment (cyb70289) [inline cpp/src/arrow/csv/writer_test.cc:181]: Move it under line 205 in delimiter test section.

**Comment by cyb70289** [comment]
Appveyor test failure looks related to this PR, needs be fixed.
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/42659260/job/qk6x3kld8ffabyuu#L2758

File: cpp/src/arrow/csv/writer_test.cc [hunk 12]
```diff
@@ -60,13 +60,15 @@ void PrintTo(const WriterTestParams& p, std::ostream* os) {
 WriteOptions DefaultTestOptions(bool include_header = false,
                                 const std::string& null_string = "",
                                 QuotingStyle quoting_style = QuotingStyle::Needed,
-                                const std::string& eol = "\n") {
+                                const std::string& eol = "\n",
+                                const char& delimiter = ',') {
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer_test.cc]: Corrected that.

File: cpp/src/arrow/csv/writer_test.cc (hunk 10 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer_test.cc]: Removed duplicate case.

File: cpp/src/arrow/csv/writer_test.cc (hunk 11 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer_test.cc]: Made the change.

**Comment by sanjibansg** [comment]
> Appveyor test failure looks related to this PR, needs be fixed. https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/42659260/job/qk6x3kld8ffabyuu#L2758

I checked the test logs. It shows an older piece of code which is failing now.
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/42659260/job/qk6x3kld8ffabyuu#L2745

Here, the error log shows this 
`assert t == read_csv(buf, read_options=read_options)`
while the [updated code](https://github.com/apache/arrow/pull/12426/files#diff-d09f577a8ebe28cba37f0339c75f9eba3bedd1dade08e10521b345667447294cR1897) has this
 ```
        assert t == read_csv(buf, read_options=read_options,
                             parse_options=parse_options)
```

**Comment by lidavidm** [comment]
Odd. I kicked the build, let's see if it persists. If it still fails maybe we can rebase and try again.

**Comment by sanjibansg** [comment]
> Odd. I kicked the build, let's see if it persists. If it still fails maybe we can rebase and try again.

@lidavidm Rebased the branch.

**Review by lidavidm — COMMENTED** [review]
LGTM.

File: cpp/src/arrow/csv/writer.cc [hunk 13]
```diff
@@ -234,7 +235,7 @@ class UnquotedColumnPopulator : public ColumnPopulator {
     while (offset < buffer_size) {
       // error happened or remaining bytes to check
       const char c = static_cast<char>(data[offset]);
-      if (c == '\n' || c == '\r' || c == ',' || c == '"') {
+      if (c == '\n' || c == '\r' || c == ',' || c == '"' || c == delimiter) {
```
Comment (pitrou) [inline cpp/src/arrow/csv/writer.cc:238]: The comma can be removed here:
```suggestion
      if (c == '\n' || c == '\r' || c == '"' || c == delimiter) {
```

File: python/pyarrow/includes/libarrow.pxd [hunk 14]
```diff
@@ -1651,7 +1651,7 @@ cdef extern from "arrow/python/csv.h" namespace "arrow::py::csv":
 cdef extern from "arrow/csv/api.h" namespace "arrow::csv" nogil:
 
     cdef cppclass CCSVParseOptions" arrow::csv::ParseOptions":
-        unsigned char delimiter
+        char delimiter
```
Comment (pitrou) [inline python/pyarrow/includes/libarrow.pxd:1654]: Character numbers in Python are unsigned, so IMHO you should keep this unsigned.

File: python/pyarrow/includes/libarrow.pxd [hunk 15]
```diff
@@ -1713,6 +1713,7 @@ cdef extern from "arrow/csv/api.h" namespace "arrow::csv" nogil:
     cdef cppclass CCSVWriteOptions" arrow::csv::WriteOptions":
         c_bool include_header
         int32_t batch_size
+        char delimiter
```
Comment (pitrou) [inline python/pyarrow/includes/libarrow.pxd:1716]: Here as well.

File: cpp/src/arrow/csv/options.cc [hunk 16]
```diff
@@ -73,6 +73,12 @@ Status ReadOptions::Validate() const {
 WriteOptions WriteOptions::Defaults() { return WriteOptions(); }
 
 Status WriteOptions::Validate() const {
+  if (ARROW_PREDICT_FALSE(delimiter == '\n' || delimiter == '\r' || delimiter == '"' ||
+                          std::string(1, delimiter) == eol)) {
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply cpp/src/arrow/csv/options.cc]: Made the change.

File: cpp/src/arrow/csv/writer.cc (hunk 13 above)
Comment (sanjibansg) [reply cpp/src/arrow/csv/writer.cc]: Made the change, removed comma from check.

File: python/pyarrow/includes/libarrow.pxd (hunk 14 above)
Comment (sanjibansg) [reply python/pyarrow/includes/libarrow.pxd]: Changed to unsigned char for delimiter.

**Comment by westonpace** [comment]
The header is still using commas instead of the delimiter (see line 524 of writer.cc).  This is causing the write_csv.py test to fail.  A more interesting question is why the Linux CI jobs are passing but I'm pretty sure that is unrelated to your PR.

**Comment by sanjibansg** [comment]
> The header is still using commas instead of the delimiter (see line 524 of writer.cc). This is causing the write_csv.py test to fail. A more interesting question is why the Linux CI jobs are passing but I'm pretty sure that is unrelated to your PR.

Yes, sorry I missed that, and previously the parse_options was missing too. Corrected that. Thanks for help!

**Review by cyb70289 — APPROVED** [review]
LGTM, thanks @sanjibansg !

**Comment by cyb70289** [comment]
"R / rhub/debian-gcc-devel:latest" job failure is not related.

**Pull request closed without merging**